### ISSUE TITLE
Correctly take orientation into account for OSI output

### ISSUE THREAD

**Issue by d-hir** [esmini/esmini#568]
As far as I saw in the code and from my own experiments, it seems that for the calculation of the OSI output, esmini assumes that the car is oriented perfectly horizontal. Any non-zero pitch and roll result in wrong positions in OSI.

OSI references all object positions by the location of their bounding box center. As esmini uses OpenSCENARIO coordinates (if I'm not mistaken) for all objects, which use the center of the rear axle projected orthogonally onto the ground plane of the object as the reference point, a coordinate transformation is needed to get the bounding box centers of the objects for OSI. This is done by taking the bounding box center offset, rotating it into the frame of the object and adding it to the OpenSCENARIO reference point.

Currently, a simplified version of exactly this procedure is done in `OSIReporter::UpdateOSIMovingObject` (**OSIReporter.cpp:860**):
```cpp
// Set OSI Moving Object Position
obj_osi_internal.mobj->mutable_base()->mutable_position()->set_x(
        objectState->state_.pos.GetX() + static_cast<double>(objectState->state_.info.boundingbox.center_.x_) * cos(objectState->state_.pos.GetH()));
obj_osi_internal.mobj->mutable_base()->mutable_position()->set_y(
        objectState->state_.pos.GetY() + static_cast<double>(objectState->state_.info.boundingbox.center_.x_) * sin(objectState->state_.pos.GetH()));
obj_osi_internal.mobj->mutable_base()->mutable_position()->set_z(
        objectState->state_.pos.GetZ() + static_cast<double>(objectState->state_.info.boundingbox.dimensions_.height_) / 2.0);
```

However, only the heading is taken into account and pitch and yaw are assumed to be zero. This is fine or at least approximately ok for most cases, when the road is completely flat. Nevertheless, as soon as there are not negligible inclinations in the road, this produces erroneous results.

The correct implementation would take the offset vector, perform the full 3D rotation and add that result to the objectState position.

**Comment by d-hir** [comment]
I don't know if this is the only place, where such an assumption is made. This is just the one that I found while debugging esmini trying to figure out why the OSI positions are not correct.

**Comment by eknabevcc** [comment]
Correct. esmini work with OpenSCENARIO reference point definition. Hence point needs to be transformed for OSI. And yes, the transformation currently only takes heading into account. Agree that this should be expanded for complete orientation. However, this is currently not top prio in our backlog. Hence, any suggested implementation, perhaps even a pull request, would be appreciated and we will prioritize merging it 👍

**Comment by jdsika** [comment]
@ClemensLinnhoff maybe you could support here?

**Comment by d-hir** [comment]
Sorry for the delay. I finally got around to testing this. From my quick tests (with the currently latest version 2.45.3) it looks like the OSI output is exactly (up to numerical precision errors) what I would expect, even on a road with quite high inclination. So thanks for fixing this!

**Comment by d-hir** [comment]
The position output seems to be fixed correctly with the changes from https://github.com/esmini/esmini/pull/607. But the velocity is still calculated using the assumption that pitch and yaw are zero. I created https://github.com/esmini/esmini/issues/685 for this.
